File: src/main/java/net/lomeli/lomlib/core/nbt/TileNBTSerialized.java
package net.lomeli.lomlib.core.nbt;

import java.lang.reflect.Field;

import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.tileentity.TileEntity;

public abstract class TileNBTSerialized extends TileEntity {
    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);
        NBTTagCompound tag = compound.getCompoundTag("tile_data_serialized");
        if (tag != null) {
            try {
                Object deserialize = NBTSerializer.fromNBT(tag, this.getClass());
                Field[] fields = this.getClass().getFields();
                if (fields != null && fields.length > 0) {
                    for (Field f : fields) {
                        if (f.getAnnotation(NBT.class) != null) {
                            Field j = deserialize.getClass().getField(f.getName());
                            if (j != null) f.set(this, j.get(deserialize));
                        }
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        super.writeToNBT(compound);
        NBTTagCompound tag = NBTSerializer.toNBT(this);
        if (tag != null) compound.setTag("tile_data_serialized", tag);
        return compound;
    }
}

File: src/main/kotlin/net/lomeli/lomlib/core/nbt/TileNBTSerialized.kt
package net.lomeli.lomlib.core.nbt

import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity

abstract class TileNBTSerialized : TileEntity() {
    override fun readFromNBT(compound: NBTTagCompound) {
        super.readFromNBT(compound)
        val tag = compound.getCompoundTag("tile_data_serialized")
        if (tag != null) {
            try {
                val deserialize = NBTSerializer.fromNBT(tag, this.javaClass)
                val fields = this.javaClass.fields
                if (fields != null && fields.size > 0) {
                    for (f in fields) {
                        if (f.getAnnotation(NBT::class.java) != null) {
                            val j = deserialize.javaClass.getField(f.name)
                            if (j != null) f.set(this, j.get(deserialize))
                        }
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun writeToNBT(compound: NBTTagCompound): NBTTagCompound {
        super.writeToNBT(compound)
        val tag = NBTSerializer.toNBT(this)
        if (tag != null) compound.setTag("tile_data_serialized", tag)
        return compound
    }
}